File: backend/src/main/java/com/readingbuddy/backend/domain/dashboard/dto/response/StageKcMasteryTrendResponse.java
package com.readingbuddy.backend.domain.dashboard.dto.response;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.List;

@Getter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class StageKcMasteryTrendResponse {

    private String stage;
    private List<KcTrend> kcTrends;

    @Getter
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class KcTrend {
        private Long kcId;
        private String kcCategory;
        private String kcDescription;
        private List<MasteryPoint> masteryTrend;
    }

    @Getter
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class MasteryPoint {
        private Float pLearn;
        private Float pTrain;
        private Float pGuess;
        private Float pSlip;
        private LocalDateTime updatedAt;
    }
}

File: reading_buddy_app/lib/features/dashboard/data/models/stage_kc_mastery_trend_response.dart
import 'package:json_annotation/json_annotation.dart';

part 'stage_kc_mastery_trend_response.g.dart';

/// Stage별 KC 숙련도 변화 추이 응답 모델
@JsonSerializable()
class StageKcMasteryTrendResponse {
  final String? stage;
  final List<KcTrend>? kcTrends;

  StageKcMasteryTrendResponse({
    this.stage,
    this.kcTrends,
  });

  factory StageKcMasteryTrendResponse.fromJson(Map<String, dynamic> json) =>
      _$StageKcMasteryTrendResponseFromJson(json);

  Map<String, dynamic> toJson() => _$StageKcMasteryTrendResponseToJson(this);
}

/// KC 추이 정보
@JsonSerializable()
class KcTrend {
  final int? kcId;
  final String? kcCategory;
  final String? kcDescription;
  final MasteryPoint? initialMastery;
  final List<MasteryPoint>? masteryTrend;

  KcTrend({
    this.kcId,
    this.kcCategory,
    this.kcDescription,
    this.initialMastery,
    this.masteryTrend,
  });

  factory KcTrend.fromJson(Map<String, dynamic> json) =>
      _$KcTrendFromJson(json);

  Map<String, dynamic> toJson() => _$KcTrendToJson(this);
}

/// 숙련도 시점 데이터
@JsonSerializable()
class MasteryPoint {
  @JsonKey(name: 'plearn')
  final double? pLearn;

  @JsonKey(name: 'ptrain')
  final double? pTrain;

  @JsonKey(name: 'pguess')
  final double? pGuess;

  @JsonKey(name: 'pslip')
  final double? pSlip;

  final DateTime? updatedAt;

  MasteryPoint({
    this.pLearn,
    this.pTrain,
    this.pGuess,
    this.pSlip,
    this.updatedAt,
  });

  factory MasteryPoint.fromJson(Map<String, dynamic> json) =>
      _$MasteryPointFromJson(json);

  Map<String, dynamic> toJson() => _$MasteryPointToJson(this);
}
